sycl/doc/design/MappingHostAddressesToDeviceEntities.md

```diff
@@ -0,0 +1,344 @@
+coverage tools, but that is not an option if a user wants to compile host part
+of an app with a 3rd-party host compiler.
+
+Further sections describe the implementation design of both approaches in more
+details, note that there are few components which should be modified regardless
+of which approach is in use.
+
+## Common front-end part
+
+DPC++ FE should support the following attribute:
+`[[__sycl_detail__::uniquely_identifiable_object(kind)]]`. This attribute accepts
+a string literal and should be applied to types (like `device_global` or
+`specialization_id`).
+
+Presence of the attribute instructs the compiler to perform the following
+things:
+- emit `sycl-unique-id` LLVM IR attribute on each definition of a variable of
+  type marked with `[[__sycl_detail__::uniquely_identifiable_object(kind)]]`
+  attribute. `sycl-unique-id` LLVM IR attribute should be accompanied by a
+  unique string identifier of a variable it is attached to. The rules for
+  creating this string are the same as for `__builtin_sycl_unique_stable_id` and
+  the same algorithm can be used when generating the string for the attribute
+- emit `sycl-uid-kind` LLVM IR attribute alongside `sycl-unique-id`, which
+  contains the `kind` string passed to
+  `[[__sycl_detail__::uniquely_identifiable_object(kind)]]` attribute
+
+**TODO**: we have `[[__sycl_detail__::device_global]]` attribute documented in
+[device global design doc][device-global-design], which instructs front-end to
+emit some additional semantic checking. Shall we leave it in place or that
+request for semantic checking should also be documented by
+`[[__sycl_detail__::uniquely_identifiable_object(kind)]]` attribute when `kind`
+is set to a certain value?
+
+**TODO**: alternatively, we could completely re-use existing
+`[[__sycl_detail__::device_global]]` attribute and introduce another one for
+specialization constants, i.e. it is a question of whether or not we want to
+generalize unique IDs generation in form of a generic attribute or not.
+
+When DPC++ compiler is used as both host and device compiler, then the attribute
+should be respected by both host and device compiler passes and LLVM IR
+attributes should appear in LLVM IR for both host and device code. When DPC++
+compiler is only used as a device compiler, then we don't expect the attribute
+to be handled on host, apparently.
+
+Another thing we need from DPC++ FE host compiler is to define a special macro,
+which will allow to distinguish it from other host compilers. That is needed to
+apply the aforementioned attribute conditionally to avoid spamming users with
+warnings about unknown attributes.
+
+**NOTE:** Alternatively we could simply set a macro which tells us whether or
+not integration footer is enabled in the compiler driver instead of creating
+a special macro for differentiating our own host compiler.
+
+The suggested macro name is `__INTEL_SYCL_HOST_COMPILER__`. It should be defined
```

Comment:
I don't know if we can use 'INTEL' in the name if we're proposing this for syclos, especially if we want to upstream this at some point. I'm also not sure if we need senior management input for using INTEL in macro name. I guess its less of a concern if its not a documented macro, but it might be better to just have a macro which indicates footer is included or something.

Reply:
I've just remembered that there is a `__SYCL_SINGLE_SOURCE__` macro documented in the [spec](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#_preprocessor_directives_and_macros), which we should use here.

AFAIK, it is not yet supported, but now it seems to be the time

Reply:
I don't think  we should use `__SYCL_SINGLE_SOURCE__` for this purpose.  The intent of that macro is to indicate a compiler that produces both host and device code in a single pass (i.e. the CFE runs just once, and then both host and device code are produced from that common IR).  This is not the case for DPC++.

Reply:
> The intent of that macro is to indicate a compiler that produces both host and device code in a single pass (i.e. the CFE runs just once, and then both host and device code are produced from that common IR).

Well, the spec says literally nothing about the number of CFE runs and based on the spec wording, DPC++ seems to be a single source compiler:

 > which produces host as well as device binary;
 
 Anyway, I'm perfectly fine with using another macro introduced by ourselves, we just need to be able to detect that we are in integration footer mode, which will imply that we are in 3rd-party host compiler mode, which will mean do not use an attribute to avoid the warning

Reply:
My understanding of spec matches @AlexeySachkov's. Unless 'single-source compiler' has standardized meaning I am unaware of, it sounds like `__SYCL_SINGLE_SOURCE__` works here 

> __SYCL_SINGLE_SOURCE__ is defined to 1 if the source file is being compiled with a SYCL single-source compiler which produces host as well as device binary;

If we can't use this, WDYT about just having a macro __INTEGRATION_FOOTER__ defined in footer.

Reply:
> If we can't use this, WDYT about just having a macro INTEGRATION_FOOTER defined in footer.

We need this macro available in regular SYCL headers, so I think that we will end up with setting something internal (`__MACRO_NAME__`) in the driver